### PR TITLE
Porting TPC Trending to QC Framework as extended Trending

### PR DESCRIPTION
**Pull request by mlesch**

### CODE REVIEW

File: Framework/src/TrendingTaskExtended.cxx [hunk 1]
```diff
@@ -0,0 +1,607 @@
...
+        mReductors[dataSource.name]->updateQuality(qo.get(), mSourcesQuality[dataSource.name]);
+        mNumberPads[dataSource.name] = 1;
+      }
+    } else {
+      ILOG(Error, Support) << "Data source '" << dataSource.type << "' unknown." << ENDM;
+    }
+  }
+
+  mTrend->Fill();
+} // void TrendingTaskExtended::trendValues(uint64_t timestamp, repository::DatabaseInterface& qcdb)
```
Comment (knopers8) [inline Framework/src/TrendingTaskExtended.cxx:127]: leftover?

File: Framework/include/QualityControl/TrendingTaskExtended.h [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+/// \brief  A extended version of the trending post-processing task.
+///
+/// A post-processing task which trends objects inside QC database (QCDB).
+/// It extracts some values of one or multiple objects using the Reductor classes,
+/// then stores them inside (a vector of) stuct "SliceInfo" which is saved in a TTree.
+/// This class extended the trending with a subrange slicer of histograms,
+/// and input/output canvas can be dealt with alongside normal histograms.
+///
+
+class TrendingTaskExtended : public PostProcessingInterface
```
Comment (knopers8) [inline Framework/include/QualityControl/TrendingTaskExtended.h:48]: How about `SliceTrendingTask`?

**Review by knopers8 — CHANGES_REQUESTED** [review]
Thanks a lot for this @mlesch! I think this is mostly good as it is, I have mostly some editorial comments about better naming. I have not tried yet to run it on my setup, but I am compiling it now and will let you know if I run into problems when executing it.

I would also like to ask you to add some documentation in this chapter: https://github.com/AliceO2Group/QualityControl/blob/master/doc/PostProcessing.md#convenience-classes
It would be nice if you could at least summarize what does the class do and how to configure it.

**Comment by knopers8** [comment]
Hi @mlesch , I played with the new task now and I am not sure if I get the results I should expect.
First of all, `ExtendedTrending_meanX_of_histogram` shows always one value for each point, even though I am running the example QC Task in parallel, so I would expect it to change. However, if I execute the ExtendedTrendingTask again, I see it picks different values. Am I doing something wrong? All the other plots always get the same values as well.

On `ExtendedTrending_meanX_of_histogram` I had a problem to see the data points actually, but this was fixed after I set `graphYRange` to none. Perhaps you could tune these parameters better in the example or leave them empty as well?

On `ExtendedTrending_meanY_of_histogram_slices` there are `slices` on the x axis, which I find a bit misleading, we still represent some parameter with a unit there, not an identifier of a slide. Perhaps the label could be more precise?

On `ExtendedTrending_meanY_of_histogram_timeMultigraph` I also had to put empty  `graphYRange`to see something, perhaps it could be left empty in the example.

I am sorry if I am nitpicking, but I believe that a clear example will be crucial to advertize the class well to other the detector teams.

File: Framework/include/QualityControl/SliceInfoTrending.h [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+
+  ClassDefNV(SliceInfo, 1);
+};
+
+struct SliceInfoQuality {
+  UInt_t qualitylevel = 0;
+  std::string title = "";
+
+  /// \brief Return the struct member/float corresponding to the argument.
+  double RetrieveValue(const std::string& varType) const
```
[the remark this answers is not in the document]
Comment (mlesch) [reply Framework/include/QualityControl/SliceInfoTrending.h]: Changed as requested

File: Framework/include/QualityControl/TrendingTaskExtended.h [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+
+#ifndef QUALITYCONTROL_TRENDINGTASKEXTENDED_H
+#define QUALITYCONTROL_TRENDINGTASKEXTENDED_H
+
+#include "QualityControl/PostProcessingInterface.h"
+#include "QualityControl/ReductorExtended.h"
+#include "QualityControl/SliceInfoTrending.h"
+#include "QualityControl/TrendingTaskExtendedConfig.h"
+
+#include <memory>
```
[the remark this answers is not in the document]
Comment (mlesch) [reply Framework/include/QualityControl/TrendingTaskExtended.h]: Changed as requested

File: Framework/include/QualityControl/ReductorExtended.h [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+#include "TAxis.h"
+
+namespace o2::quality_control::postprocessing
+{
+/// \brief An interface for storing data derived from QC objects into a TTree.
+///
+/// A extended reductor class from which each reductor used for the extended trending.
+///
+
+class ReductorExtended
```
[the remark this answers is not in the document]
Comment (mlesch) [reply Framework/include/QualityControl/ReductorExtended.h]: Changed as requested

File: Modules/Common/include/Common/QualityReductorExtended.h [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+using namespace o2::quality_control::postprocessing;
+namespace o2::quality_control_modules::common
+{
+/// \brief A reductor of quality objects for the extended trending.
+///
+/// A Reductor for the quality objects. It receives
+/// a struct of 'SliceInfoQuality'
+///
+
+class QualityReductorExtended : public quality_control::postprocessing::ReductorExtended
```
[the remark this answers is not in the document]
Comment (mlesch) [reply Modules/Common/include/Common/QualityReductorExtended.h]: Removed as requested as it does not bring anything new compared to the framework trending

File: Framework/src/TrendingTaskExtended.cxx (hunk 1 above)
Comment (mlesch) [reply Framework/src/TrendingTaskExtended.cxx]: Changed as requested

File: Framework/include/QualityControl/TrendingTaskExtended.h (hunk 2 above)
Comment (mlesch) [reply Framework/include/QualityControl/TrendingTaskExtended.h]: Changed as requested

**Comment by mlesch** [comment]
Hi @knopers8 , 
I implemented the requested changes and fixed a bug. Now the trending changes over time when running the QC Task in parallel. I also tried to precise the labels of the plots in the json file a bit better, but of course I can change them if the current version is not clear enough. I will now start working the documentation about the class.

**Review by knopers8 — APPROVED** [review]

**Pull request merged**